### PR TITLE
Partial fix #2923: Improve javalib 64 bit UnixProcess waitFor handling

### LINKED ISSUE #2923: ProcessTest is earning its keep by failing intermittently, especially on macOS.
I have noticed that `ProcessTest` is failing intermittently recently.  This seems to be
happening more often on macOS.  I am not sure if this started happening before
or after my recent (few weeks) `posix_spawn` changes were merged.

As frustrating as intermittent ProceesTest failures are, I suspect that the
test is validly picking up a hard-to-provoke defect in or releated to 
javalib Process.scala.

The current failures seem to be with Process trying to read from an 
InputStream, to which the/a child process is expected to have written.

The next-most recent failures tended to be on Windows and had
to do with it taking a long time (> 10 or 20 or more seconds) for
a process to exit.

Here, I think there is a race condition and/or an I/O is getting lost.
(because parent is closing wrong parts of the involved pipes?
because the child exits aborning or never exists at all?)

There may also be something happening with the whole ProcessMonitor thing.

On the whole, I think there is a real issue here. I will investigate, but wanted
others to know not to blame poor ProcessTest for sounding the alarm, albeit
in hard to understand ways.

### PR DESCRIPTION
**Pull request by LeeTibbert**
The goal of this PR is to reduce the defects which ProcessTest is validly reporting in Continuous Integration (CI).
This has been seen particularly often on macOS.
It is marked "Partial fix" because many hundreds of CI runs need to be monitored to see if it has the desired
effect. It is hard to prove a negative.

This PR has benefited from conversations with Arman Bilge and study of his `armanbilge/epollcat` repository,
particularly the `kevent64` handling. Merit may go to Arman; I retain ownership of all bugs & defects.

This PR makes changes in javalib for 64 bit systems running either Linux >= 5.3 or macOS. 
All 32 bit systems, older Linux systems, and systems running other operating systems
continue to use the now legacy `UnixProcess/UnixProcessGen1` code.

When/if it proves itself, it could probably be extended to FreeBSD and other BSD derivative.

Reviewers will note that the new `UnixProcesGen2` code does not use the C++ `ProcessMonitor` code.

This PR provides the basis for a second PR which will modify `UnixProcessGen2` to use `posix_spawn()`.
That method is faster than `fork()` and a step on the road to using an even faster upcoming `uring_spawn()`.
Pass the hot sauce please.

### CODE REVIEW

File: javalib/src/main/scala/java/lang/process/UnixProcessGen2.scala [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+  override def waitFor(timeout: scala.Long, unit: TimeUnit): scala.Boolean = {
+    if (_exitValue.isDefined) { // avoid wait-after-wait complexity
+      true
+    } else {
+      val tv = stackalloc[timespec]()
+      fillTimeval(timeout, unit, tv)
+
+      // wait until process exits or times out.
+      osWaitForImpl(Some(tv)).isDefined
+    }
```
Comment (WojciechMazur) [inline javalib/src/main/scala/java/lang/process/UnixProcessGen2.scala:104]:
```suggestion
    _exitValue // avoid wait-after-wait complexity
     .orElse {
      // wait until process exits or times out.
      val tv = stackalloc[timespec]()
      fillTimeval(timeout, unit, tv)
      osWaitForImpl(Some(tv))
    }.isDefined
```

**Review by WojciechMazur — COMMENTED** [review]
Looks good, thank you for your work in improving work with processes. 
Overall new algorithm looks good to me, I belive you have far better knowledge of working with poll events then I do. I only have few suggestions for working with Options to make code a bit more readable and idiomatic.

File: javalib/src/main/scala/java/lang/process/UnixProcessGen2.scala [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+  override def getOutputStream(): OutputStream = _outputStream
+
+  override def isAlive(): scala.Boolean = {
+    waitpidImpl(pid, options = WNOHANG) == 0
+  }
+
+  override def toString = { // Match JVM output
+    val ev =
+      if (_exitValue.isEmpty) "not exited"
+      else _exitValue.head.toString()
```
[the remark this answers is not in the document]
Comment (LeeTibbert) [reply javalib/src/main/scala/java/lang/process/UnixProcessGen2.scala:82]: Done. Thank you for the improvement.

File: javalib/src/main/scala/java/lang/process/UnixProcessGen2.scala [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+
+    s"Process[pid=${pid}, exitValue=${ev}]"
+  }
+
+  override def waitFor(): scala.Int = {
+    if (_exitValue.isDefined) { // avoid wait-after-wait complexity
+      _exitValue.head
+    } else { // wait until process exits or forever, whichever comes first.
+      osWaitForImpl(None).getOrElse(1) // 1 == EXIT_FAILURE, unknown cause
+    }
```
[the remark this answers is not in the document]
Comment (LeeTibbert) [reply javalib/src/main/scala/java/lang/process/UnixProcessGen2.scala:92]: Done

File: javalib/src/main/scala/java/lang/process/UnixProcessGen2.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (LeeTibbert) [reply javalib/src/main/scala/java/lang/process/UnixProcessGen2.scala]: > Overall new algorithm looks good to me, I belive you have far better knowledge of working with poll events then I do
Thank you for the encouragement, especially  coming from the person who re-wrote the unix Process and waitFor for
Windows.

**Review by WojciechMazur — APPROVED** [review]

**Comment by LeeTibbert** [comment]
Thank you for the merge.

After this PR has proved itself in the 0.5.0 stream for a while (weeks?), I believe
it is a candidate for the 0.4.n stream.  Let's see how the CI logs go.

**Pull request merged**